File: week-11/src/services/analyticsService.js
// Analytics Service for Mixpanel Integration
// Handles all event tracking for the application

class AnalyticsService {
  constructor() {
    this.mixpanelToken = process.env.REACT_APP_MIXPANEL_TOKEN;
    this.enabled = !!this.mixpanelToken && process.env.NODE_ENV === 'production';
    
    if (this.enabled && window.mixpanel) {
      window.mixpanel.init(this.mixpanelToken);
    }
  }

  // Track generic event
  trackEvent(eventName, properties = {}) {
    if (!this.enabled || !window.mixpanel) {
      console.log('Analytics event:', eventName, properties);
      return;
    }

    try {
      const eventProperties = {
        ...properties,
        timestamp: new Date().toISOString(),
        environment: process.env.NODE_ENV
      };

      window.mixpanel.track(eventName, eventProperties);
    } catch (error) {
      console.error('Error tracking event:', error);
    }
  }

  // Track page view
  trackPageView(pageName, properties = {}) {
    this.trackEvent('page_view', {
      ...properties,
      page_name: page_name
    });
  }

  // Track button click
  trackButtonClick(buttonName, properties = {}) {
    this.trackEvent('button_click', {
      ...properties,
      button_name: buttonName
    });
  }

  // Track modal interaction
  trackModalInteraction(user, modalName, action, properties = {}) {
    this.trackEvent('modal_interaction', {
      ...properties,
      modal_name: modalName,
      action: action,
      user_id: user?.id,
      user_type: user?.user_type
    });
  }

  // Track flagging action
  trackFlagging(user, flagType, properties = {}) {
    this.trackEvent('flagging_action', {
      ...properties,
      flag_type: flagType,
      user_id: user?.id,
      user_type: user?.user_type
    });
  }

  // Track policy acknowledgment
  trackPolicyAcknowledgment(user, policyId, properties = {}) {
    this.trackEvent('policy_acknowledged', {
      ...properties,
      policy_id: policyId,
      user_id: user?.id,
      user_type: user?.user_type
    });
  }

  // Track save button interaction
  trackSaveButton(user, action, entityType, entityId, properties = {}) {
    this.trackEvent('save_button_interaction', {
      ...properties,
      action: action, // 'saved' or 'unsaved'
      entity_type: entityType,
      entity_id: entityId,
      user_id: user?.id,
      user_type: user?.user_type
    });
  }

  // Identify user
  identify(user) {
    if (!this.enabled || !window.mixpanel) return;

    try {
      window.mixpanel.identify(user.id.toString());
      window.mixpanel.people.set({
        '$email': user.email,
        '$name': user.full_name,
        'user_type': user.user_type,
        'created_at': user.created_at
      });
    } catch (error) {
      console.error('Error identifying user:', error);
    }
  }

  // Set user properties
  setUserProperties(properties) {
    if (!this.enabled || !window.mixpanel) return;

    try {
      window.mixpanel.people.set(properties);
    } catch (error) {
      console.error('Error setting user properties:', error);
    }
  }
}

// Export singleton instance
export default new AnalyticsService();

// Export individual functions for convenience
export const trackEvent = (eventName, properties) => {
  analyticsService.trackEvent(eventName, properties);
};

export const trackPageView = (pageName, properties) => {
  analyticsService.trackPageView(pageName, properties);
};

export const trackButtonClick = (buttonName, properties) => {
  analyticsService.trackButtonClick(buttonName, properties);
};

export const trackModalInteraction = (user, modalName, action, properties) => {
  analyticsService.trackModalInteraction(user, modalName, action, properties);
};

export const trackFlagging = (user, flagType, properties) => {
  analyticsService.trackFlagging(user, flagType, properties);
};

export const trackPolicyAcknowledgment = (user, policyId, properties) => {
  analyticsService.trackPolicyAcknowledgment(user, policyId, properties);
};

export const trackSaveButton = (user, action, entityType, entityId, properties) => {
  analyticsService.trackSaveButton(user, action, entityType, entityId, properties);
};

export const identifyUser = (user) => {
  analyticsService.identify(user);
};

File: week-11/app/services/analytics_service.rb
# Analytics Service
# Handles Mixpanel event tracking

class AnalyticsService
  class << self
    # Track event to Mixpanel
    def track_event(user, event_name, properties = {})
      return unless mixpanel_enabled?

      begin
        tracker = Mixpanel::Tracker.new(ENV['MIXPANEL_TOKEN'])

        # Add user properties
        user_properties = {
          user_id: user.id,
          email: user.email,
          user_type: user.user_type,
          created_at: user.created_at
        }

        # Merge with custom properties
        event_properties = user_properties.merge(properties).merge(
          timestamp: Time.current.to_i,
          environment: Rails.env
        )

        # Track event
        tracker.track(user.id.to_s, event_name, event_properties)

        # Update user profile
        tracker.people.set(user.id.to_s, user_properties)

        Rails.logger.info "Analytics event tracked: #{event_name} for user #{user.id}"
      rescue => e
        Rails.logger.error "Error tracking analytics event: #{e.message}"
        # Don't raise error, analytics failures shouldn't break the app
      end
    end

    # Track page view
    def track_page_view(user, page_name, properties = {})
      track_event(user, 'page_view', properties.merge(page_name: page_name))
    end

    # Track button click
    def track_button_click(user, button_name, properties = {})
      track_event(user, 'button_click', properties.merge(button_name: button_name))
    end

    # Track modal interaction
    def track_modal_interaction(user, modal_name, action, properties = {})
      track_event(user, 'modal_interaction', properties.merge(
        modal_name: modal_name,
        action: action
      ))
    end

    # Track flagging action
    def track_flagging(user, flag_type, properties = {})
      track_event(user, 'flagging_action', properties.merge(flag_type: flag_type))
    end

    # Track policy acknowledgment
    def track_policy_acknowledgment(user, policy_id, properties = {})
      track_event(user, 'policy_acknowledged', properties.merge(policy_id: policy_id))
    end

    # Track save button interaction
    def track_save_button(user, action, entity_type, entity_id, properties = {})
      track_event(user, 'save_button_interaction', properties.merge(
        action: action, # 'saved' or 'unsaved'
        entity_type: entity_type,
        entity_id: entity_id
      ))
    end

    private

    def mixpanel_enabled?
      ENV['MIXPANEL_TOKEN'].present? && Rails.env.production?
    end
  end
end

